core/consensus/babe/src/lib.rs

```diff
@@ -281,44 +292,40 @@ impl<Hash, H, B, C, E, I, Error, SO> SlotWorker<B> for BabeWorker<C, E, I, SO> w
 			return Box::new(future::ok(()));
 		}
 
-		// FIXME replace the dummy empty slices with real data
-		// https://github.com/paritytech/substrate/issues/2435
-		// https://github.com/paritytech/substrate/issues/2436
-		let proposal_work = if let Some(((inout, proof, _batchable_proof), index)) = claim_slot(
-			&[0u8; 0],
+		let proposal_work = if let Some(((inout, vrf_proof, _batchable_proof), authority_index)) = claim_slot(
+			&randomness,
 			slot_info.number,
-			&[0u8; 0],
-			0,
-			&authorities,
+			slot_info.number / self.slots_per_epoch,
```

Comment:
I am fairly certain I have warned against this bug already in a couple places, but in case it wasn't clear: this calculation is wrong in the case where there have been more than `slots_per_epoch` slots since the parent block.

In this part of the code we should always draw the epoch number and randomness from the state of the best block (but note that we are then _committing_ to build on top of this block _before_ deciding whether we will build). This is safe because the head of the chain has always been imported and can be assumed to have available state. Authoring would be impossible otherwise.

Reply:
Fixed, by moving epoch calculation entirely to the runtime.